object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/ObjectConstructionAnnotatedTypeFactory.java

```diff
@@ -279,6 +306,276 @@ private static String getExactStringValue(Tree tree, ValueAnnotatedTypeFactory f
     return null;
   }
 
+  @Override
+  public void postAnalyze(ControlFlowGraph cfg) {
+    checkCFG(cfg.getEntryBlock(), cfg);
+    super.postAnalyze(cfg);
+  }
+
+  private void checkCFG(SpecialBlock entry, ControlFlowGraph cfg) {
+
+    Set<Pair<Long, Set<AssignmentNode>>> visited = new HashSet<>();
+    visited.add(Pair.of(entry.getId(), Collections.emptySet()));
+
+    Deque<Pair<BlockImpl, Set<AssignmentNode>>> worklist = new ArrayDeque<>();
+    worklist.add(Pair.of((BlockImpl) entry, Collections.emptySet()));
+
+    while (!worklist.isEmpty()) {
+
+      Pair<BlockImpl, Set<AssignmentNode>> state = worklist.getLast();
+      worklist.removeLast();
+      List<Node> nodes = getBlockNodes(state.first);
+      Set<AssignmentNode> newDefs = new HashSet<>(state.second);
+
+      for (Node node : nodes) {
+        if (node instanceof AssignmentNode) {
+          Node lhs = ((AssignmentNode) node).getTarget();
+          Node rhs = ((AssignmentNode) node).getExpression();
+
+          if (lhs instanceof LocalVariableNode && hasAlwaysCall(lhs.getType())) {
+
+            // Reassignment to the lhs
+            if (isVarInPrevDefs(newDefs, (LocalVariableNode) lhs)) {
```

Comment:
Why pass `newDefs` here?  Shouldn't you pass `state.second`?

Reply:
I think it should be newDefs, consider the following test case:
void ownershipTransfer(){
        Foo f1 = new Foo();
        Foo f2 = f1;
   }
the state is Pair(RegularBlock([f1 = new Foo(), f2, f1, f2 = f1,]), []), so state.second is empty list, but newDefs contains f1 = new Foo() after the first node. Then, the result of isVarInPrevDefs(state.second, rhs) in the following check is different from isVarInPrevDefs(newDefs, rhs):
if ((rhs instanceof ObjectCreationNode)
                || (rhs instanceof MethodInvocationNode)
                || (rhs instanceof LocalVariableNode
                    && isVarInPrevDefs(newDefs, (LocalVariableNode) rhs))){
          newDefs.add((AssignmentNode) node);
}

Reply:
I think it should be newDefs, consider the following example:
    void ownershipTransfer(){
        Foo f1 = new Foo();
        Foo f2 = new Foo();
        f2 = f1;
        f2.a();
    }
The state is Pair(RegularBlock([f2 = new Foo(), f2, f1, f2 = f1, f2]), [f1 = new Foo()]) when we are checking f2 = f1, so state.second does not contain f2 = new Foo(), but newDefs contains {f1 = new Foo(), f2 = new Foo()}. Thus, if we use state.second instead of newDefs, then isVarInPrevDefs(state.second, (LocalVariableNode) lhs) returns false and it won't report error for Foo f2 = new Foo();

Reply:
Ok, I see.  My pseudocode assumed a single statement per block.  Here, I think we should change the code and have `prevDefs` and `newDefs` available in the loop over nodes in a block.  At the beginning of each loop iteration, `newDefs` should be a new Set containing `prevDefs`, and at the end of each loop iteration, we should have `prevDefs = newDefs`.  Then, all calls to `isVarInPrevDefs` should pass `prevDefs`.  Does that make sense?

Reply:
Why should we change newDefs and use prevDefs? do you have any example to show the difference?

Reply:
I think now that the code is correct as-is.  But let's rename the method `isVarInPrevDefs` to just `isVarInDefs` to avoid confusion.